azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import abc
+import six
+from collections import deque
+from . import pipeline_ops
+from . import pipeline_events
+from . import mqtt_topic_helper
+from azure.iot.device.iothub.models import Message
+from azure.iot.device.common.transport.mqtt.mqtt_provider import MQTTProvider
+from azure.iot.device.common.transport import constant
+
+logger = logging.getLogger(__name__)
+
+# TODO: standardize on {} instead of %s
+# TODO: standardize logs with {}({}) prefix
+
+
+@six.add_metaclass(abc.ABCMeta)
+class PipelineStage(object):
+    """
+    Base class representing a stage in the processing pipeline.  Each stage is responsible for receiving
+    PipelineOperation objects from the top, possibly processing them, and possibly passing them down.  It
+    is also responsible for receiving PipelineEvent objects from the bottom, possibly processing them, and
+    possibly passing them up.  In both cases (PipelineOperation and PipelineEvent), the default behavior
+    should be to pass along an operation or event that you don't handle.
+    """
+
+    def __init__(self):
+        self.name = self.__class__.__name__
+        self.next = None
+        self.previous = None
+        self.pipeline_root = None
+
+    def run_op(self, op):
+        """
+        Run the given operation.  This is the public function that outside callers would call to run an
+        operation.  Derived classes should override the _run_op function to impliment stage-specific behavior.
+        """
+        try:
+            self._run_op(op)
+        except Exception as e:
+            logger.error(msg="Error in _run_op call", exc_info=e)
+            op.error = e
+            self.complete_op(op)
+
+    def run_ops_serial(self, *args, **kwargs):
+        """
+        Run the operations passed in *args in a serial manner, such that each operation waits for the
+        previous operation to be completed before running.  If an operation fails, the operations that
+        follow will not be run, but the operation passed using the finally_op will be run.  After
+        all operations and the finally_op (if provided) are run, then the function passed in the
+        callback parameter will be called, passing the last operation (in the event of success)
+        or the failing operation (in the event of failure).
+        """
+        worklist = deque(args)
```

Comment:
>deque(args) [](start = 19, length = 11)

Interested - was there a particular reason you used a deque and it's .popleft method instead of using a list, and it's .pop method?

Reply:
No technical reason.  It was used in some code I was reading and I was curious to use it.  Any reason to change it to a list?

---
In reply to: [279654358](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279654358) [](ancestors = 279654358)

Reply:
OK, I found the reason: [https://stackoverflow.com/questions/21639888/is-it-possible-to-convert-list-to-queue-in-python](https://stackoverflow.com/questions/21639888/is-it-possible-to-convert-list-to-queue-in-python)
popleft is O(n) for list of size n.  not a big deal here since our lists are small.
making a Queue from a list is slow (probably O(n)) compared to making a deque from a list (>300 times faster according to some comment in the stackoverflow discussion).
I think I'll just use Queue, but the efficiency tradeoff is still interesting.

---
In reply to: [279901078](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279901078) [](ancestors = 279901078,279654358)